### PR TITLE
Make experimental a runtime flag

### LINKED ISSUE #26713: Changing the definition of experimental
Since 1.12.0, experimental is not a nightly build anymore but is a different channel that follows the same release scheduled than stable. As it stands now, experimental is just a super-set of the stable release, and should not be considered "beta".

**TL;DR:** experimental _is_ a stable release, with a couple features which API stability we don't guarantee.

## Proposal

- [x] Change experimental to be a runtime flag (daemon configuration, environment variable, ...?) rather than a compile-time build tag.
- [ ] Allow the client to discover whether the daemon it's talking to has experimental enabled, and display the extra experimental commands/flags accordingly. Note that this extra round-trip will likely be required to solve #25498 anyways.
- [x] Remove traces of experimental in build scripts (e.g., `DOCKER_EXPERIMENTAL` environment variable).
- [ ] Find a transition path for https://experimental.docker.com (e.g., should it install a regular daemon with the experimental flag enabled?).

### PR DESCRIPTION
**Pull request by mlaventure**
Signed-off-by: Kenfe-Mickael Laventure mickael.laventure@gmail.com
## 

Left to do:
- [ ] Update the install script to automatically set `"experimental": true` if fetched from `https://experimental.docker.com/` 
- [x] ~~See what solution is adopted by #25498 and piggy back on it to check the daemon experimental flag~~
- [x] Update documentation

Closes #26713

### CODE REVIEW

File: integration-cli/check_test.go [hunk 1]
```diff
@@ -264,7 +264,11 @@ func (s *DockerSwarmSuite) AddDaemon(c *check.C, joinSwarm, manager bool) *Swarm
 		port:   defaultSwarmPort + s.portIndex,
 	}
 	d.listenAddr = fmt.Sprintf("0.0.0.0:%d", d.port)
-	err := d.StartWithBusybox("--iptables=false", "--swarm-default-advertise-addr=lo") // avoid networking conflicts
+	args := []string{"--iptables=false", "--swarm-default-advertise-addr=lo"}
+	if experimentalDaemon {
+		args = append(args, "--experimental")
+	}
+	err := d.StartWithBusybox(args...) // avoid networking conflicts
```
Comment (mlaventure) [inline integration-cli/check_test.go]: Updated, thanks.

**Comment by icecrime** [comment]
> See what solution is adopted by #25498 and piggy back on it to check the daemon experimental flag

I think it's perfectly fine to rely on `/info` in the meantime (we're just talking about optimizations here).

Otherwise design LGTM :+1:

**Comment by icecrime** [comment]
> See what solution is adopted by #25498 and piggy back on it to check the daemon experimental flag

This is a requirement, but it won't change much about this PR, so I'm moving it to code review so we can try and get it in for 1.13.0 :+1:

**Comment by justincormack** [comment]
This seems to be working as expected, code looks fine.

LGTM

**Comment by LK4D4** [comment]
LGTM
@mlaventure but need rebase

**Comment by thaJeztah** [comment]
We should have a CI check or label to prevent future PR's from adding an experimental build-flag

/cc @ehazlett @mlaventure

File: integration-cli/docker_experimental_network_test.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// +build experimental
+// +build !windows
```
Comment (tianon) [inline integration-cli/docker_experimental_network_test.go]: Were we just never running tests on `windows` with `experimental` enabled previously?

File: hack/make/.integration-daemon-start [hunk 3]
```diff
@@ -50,6 +50,11 @@ if [ "$DOCKER_REMAP_ROOT" ]; then
 	extra_params="--userns-remap $DOCKER_REMAP_ROOT"
 fi
 
+if [  "$DOCKER_EXPERIMENTAL" ]; then
+	echo >&2 '# DOCKER_EXPERIMENTAL is set: starting daemon with experimental features enabled! '
+	extra_params="$extra_params --experimental"
+fi
```
Comment (tianon) [inline hack/make/.integration-daemon-start]: Aren't the experimental features tested regardless of whether this is set on our integration daemon, or does this mean we'll still need to run the full test suite twice (once without and once with this set)?

File: hack/make/build-rpm [hunk 4]
```diff
@@ -96,8 +96,7 @@ set -e
 			# Install runc and containerd
 			RUN ./hack/dockerfile/install-binaries.sh runc-dynamic containerd-dynamic
 		EOF
-
-		if [ "$DOCKER_EXPERIMENTAL" ]; then
+		if [[ "$VERSION" == *-dev ]] || [ -n "$(git status --porcelain)" ]; then
 			echo 'ENV DOCKER_EXPERIMENTAL 1' >> "$DEST/$version/Dockerfile.build"
 		fi
```
Comment (mlaventure) [inline hack/make/build-rpm]: Should I remove [`--define '_experimental ${DOCKER_EXPERIMENTAL:-0}'`](https://github.com/docker/docker/blob/2c620d0aa24c5f774a9115449a86b158b005bba8/hack/make/build-rpm#L119) too ?

File: integration-cli/docker_experimental_network_test.go (hunk 2 above)
Comment (mlaventure) [inline integration-cli/docker_experimental_network_test.go]: Correct, all the test have `DaemonIsLinux` as a prerequisite

File: hack/make/.integration-daemon-start (hunk 3 above)
Comment (mlaventure) [inline hack/make/.integration-daemon-start]: Without it the experimental endpoint are not "exposed" to the clients.

However, if there''s code elsewhere that is supposed to be experimental and isn't guard by the proper check, testing both with and without the flag should help catch it.

**Comment by mlaventure** [comment]
- Added a new `Docker-Experimental` header to every request.
- Using `/_ping` endpoint to check for the above value to determine if experimental is enabled on the daemon.

**Comment by mlaventure** [comment]
@thaJeztah I couldn't find a place where to document the extra header, any pointers?

**Comment by thaJeztah** [comment]
@mlaventure I'd consider adding it to the API changes at least; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v125-api-changes

Perhaps it should be in the introduction, just like we mention error messages (i.e. explain which headers are returned by API calls) https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.25.md#1-brief-introduction 

@mstanleyjones wdyt?

**Comment by mlaventure** [comment]
@thaJeztah added it to the api changes.

For the rest, it can be done on a separate PR. I wouldn't even know where to start :sweat_smile:

**Comment by mlaventure** [comment]
ping @LK4D4 @justincormack one last review?

I'll do the script changes in a separate PR, the script has to be manually tested anyhow.

**Comment by mlaventure** [comment]
Both experimental & janky are actually :green_heart:

File: api/server/middleware/experimental.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"net/http"
+
+	"golang.org/x/net/context"
+)
+
+// ExperimentalMiddleware is a the middleware in charge of adding the
+// 'Docker-Experimental' header to every outgoing request
+type ExperimentalMiddleware struct {
```
Comment (vdemeester) [inline api/server/middleware/experimental.go]: Doesn't `golint` errors out on this :sweat_smile: ? `middleware.ExperimentalMiddleware` seams a bit redondant :angel: :stuck_out_tongue_closed_eyes:

File: docs/reference/api/docker_remote_api.md [hunk 6]
```diff
@@ -146,7 +146,7 @@ This section lists each version from latest to oldest.  Each listing includes a
 * `POST /containers/prune` prunes stopped containers.
 * `POST /images/prune` prunes unused images.
 * `POST /volumes/prune` prunes unused volumes.
-
+* Every API request answer will now include a `Docker-Experimental` specifying if experimental feature are enabled (value can be `true` or `false`).
```
Comment (vdemeester) [inline docs/reference/api/docker_remote_api.md]: It's not docs-review yet but shouldn't this be `Every API response …` ?

**Review by vdemeester — COMMENTED** [review]
Few comments, but looks good :angel:

File: docs/reference/api/docker_remote_api.md (hunk 6 above)
Comment (mlaventure) [inline docs/reference/api/docker_remote_api.md]: Yep, sounds better, thanks!

File: api/server/middleware/experimental.go (hunk 5 above)
Comment (vdemeester) [inline api/server/middleware/experimental.go]: ah, ok :angel: nevermind then :stuck_out_tongue_closed_eyes: We could (should) rename those I think, but it's not the scope of the PR 👼

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Comment by vdemeester** [comment]
@thaJeztah @mlaventure should we move this to docs-review or should we do that in a follow up PR?

**Review by vieux — APPROVED** [review]
LGTM

**Comment by ehazlett** [comment]
@docker/core-machine-maintainers  machine will need some updates to support the new experimental flag.

LGTM

**Comment by thaJeztah** [comment]
Had one small nit on docs, but feel free to go ahead if you prefer

**Comment by mlaventure** [comment]
@thaJeztah made the doc change.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
all green!

**Comment by tianon** [comment]
https://github.com/docker/docker/blob/master/experimental/README.md should probably be updated for this change as well, right?

**Comment by thaJeztah** [comment]
@tianon see https://github.com/docker/docker/issues/28160 :heart:

**Comment by praving55** [comment]
Can you point me to the reference please? I could not see it in docs

**Comment by thaJeztah** [comment]
docs for 1.13 are not yet published, but the flag itself is documented in the `dockerd` reference. You can either set a `--experimental` flag on the daemon (but depends on what init system (`upstart`, `sysv-init`, `systemd`) you're using, or use a `daemon.json` configuration file, as described here; https://github.com/docker/docker/blob/v1.13.0/docs/reference/commandline/dockerd.md#daemon-configuration-file

**Comment by praving55** [comment]
Is there a reason docs are lagging? I never saw that happening before.

**Comment by thaJeztah** [comment]
Not all parts of the release are released yet, so the release will be officially announced after all projects have finished their release, and the docs will probably go out together with that.

**Comment by praving55** [comment]
ok

**Pull request merged**